torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -1210,6 +1211,12 @@ class AsyncSparseAllreduceCUDAWork : public AsyncSparseAllreduceWork {
       guard.set_index(inputs[i].device().index());
       events[i].block(at::cuda::getCurrentCUDAStream());
     }
+
+   // Copy outputs back to inputs after synchronization, so that users can 
+   // access all reduce results from input tensors
+   for (size_t i = 0; i < inputs.size(); i++) {
+      inputs[i].copy_(outputs[i]);
```

Comment:
Let's add a TODO here, this can potentially be non_blocking if we do pinned_memory correctly.

Reply:
it is device to device copy, I think we do not need non_blocking here?

Reply:
I see. Yep, D2D copy will be synced on current streams on both devices.